src/Microsoft.SqlTools.ServiceLayer/FileBrowser/FileBrowserService.cs

```diff
@@ -0,0 +1,222 @@
+                }
+                return connectionService;
+            }
+            set
+            {
+                connectionService = value;
+            }
+        }
+
+        /// <summary>
+        /// Initializes the service instance
+        /// </summary>
+        /// <param name="serviceHost"></param>
+        /// <param name="context"></param>
+        public void InitializeService(ServiceHost serviceHost)
+        {
+            // Open a file browser
+            serviceHost.SetRequestHandler(FileBrowserOpenRequest.Type, HandleFileBrowserOpenRequest);
+
+            // Expand a folder node
+            serviceHost.SetRequestHandler(FileBrowserExpandRequest.Type, HandleFileBrowserExpandRequest);
+
+            // Validate the selected files
+            serviceHost.SetRequestHandler(FileBrowserValidateRequest.Type, HandleFileBrowserValidateRequest);
+
+            // Close the file browser
+            serviceHost.SetRequestHandler(FileBrowserCloseRequest.Type, HandleFileBrowserCloseRequest);
+        }
+
+        internal async Task HandleFileBrowserOpenRequest(
+            FileBrowserOpenParams fileBrowserParams,
+            RequestContext<FileBrowserOpenResponse> requestContext)
+        {
+            FileBrowserOpenResponse response = new FileBrowserOpenResponse();
+
+            try
+            {
+                ConnectionInfo connInfo;
+                this.ConnectionServiceInstance.TryFindConnection(
+                    fileBrowserParams.OwnerUri,
+                    out connInfo);
+
+                if (connInfo != null)
+                {
+                    SqlConnection sqlConn = GetSqlConnection(connInfo);
+                    FileBrowserOperation browser = new FileBrowserOperation(sqlConn, fileBrowserParams.ExpandPath, fileBrowserParams.FileFilters);
+                    browser.PopulateFileTree();
+
+                    // Cache the file browser operation for expanding node request in the future
+                    this.ownerToFileBrowserMap.Add(fileBrowserParams.OwnerUri, browser);
+
+                    response.FileTree = browser.FileTree;
+                    response.Succeeded = true;
+                }
+            }
+            catch (Exception ex)
+            {
+                response.Succeeded = false;
+                response.Message = ex.Message;
+            }
+
+            await requestContext.SendResult(response);
+        }
+
+        internal async Task HandleFileBrowserExpandRequest(
+            FileBrowserExpandParams fileBrowserParams,
+            RequestContext<FileBrowserExpandResponse> requestContext)
+        {
+            FileBrowserExpandResponse response = new FileBrowserExpandResponse();
+
+            try
+            {
+                FileBrowserOperation browser = this.ownerToFileBrowserMap[fileBrowserParams.OwnerUri];
+                if (browser != null)
+                {
+                    browser.ExpandSelectedNode(fileBrowserParams.ExpandPath);
```

Comment:
>ExpandSelectedNode [](start = 28, length = 18)

similar comment here?  should these methods be async and send a notification when complete to avoid blocking the dispatcher?

Reply:
changed to async